File: frontend/src/utils/validation.js
/**
 * Client-side validation utilities
 */

const VALID_TYPES = ['image/png', 'image/jpeg', 'application/pdf']
const MAX_FILE_SIZE = 10 * 1024 * 1024 // 10MB
const MIN_FILE_SIZE = 10 * 1024 // 10KB

/**
 * Validate blueprint file before upload
 * @param {File} file - File to validate
 * @returns {Array<Object>} Array of validation errors
 */
export function validateBlueprint(file) {
  const errors = []

  if (!file) {
    errors.push({
      code: 'NO_FILE',
      message: 'No file selected',
      severity: 'error'
    })
    return errors
  }

  // Check file type
  if (!VALID_TYPES.includes(file.type)) {
    errors.push({
      code: 'INVALID_TYPE',
      message: 'Please upload PNG, JPG, or PDF files only',
      severity: 'error'
    })
  }

  // Check file size
  if (file.size > MAX_FILE_SIZE) {
    errors.push({
      code: 'FILE_TOO_LARGE',
      message: `File size (${(file.size / 1024 / 1024).toFixed(1)}MB) exceeds 10MB limit`,
      severity: 'error'
    })
  }

  if (file.size < MIN_FILE_SIZE) {
    errors.push({
      code: 'FILE_TOO_SMALL',
      message: 'File appears to be too small or corrupted',
      severity: 'warning'
    })
  }

  return errors
}

/**
 * Check if file is a valid image by checking magic bytes
 * @param {File} file - File to check
 * @returns {Promise<boolean>} True if valid image format
 */
export async function isValidImageFormat(file) {
  return new Promise((resolve) => {
    const reader = new FileReader()

    reader.onload = (e) => {
      const arr = new Uint8Array(e.target.result)

      // Check PNG signature
      if (arr.length >= 8 &&
          arr[0] === 0x89 && arr[1] === 0x50 &&
          arr[2] === 0x4E && arr[3] === 0x47) {
        resolve(true)
        return
      }

      // Check JPEG signature
      if (arr.length >= 2 &&
          arr[0] === 0xFF && arr[1] === 0xD8) {
        resolve(true)
        return
      }

      // Check PDF signature
      if (arr.length >= 4 &&
          arr[0] === 0x25 && arr[1] === 0x50 &&
          arr[2] === 0x44 && arr[3] === 0x46) {
        resolve(true)
        return
      }

      resolve(false)
    }

    reader.onerror = () => resolve(false)

    reader.readAsArrayBuffer(file.slice(0, 8))
  })
}

File: backend/utils/validation.py
"""
Input Validation
Validates blueprint image data before processing
"""

import logging

logger = logging.getLogger(__name__)

# Validation configuration
MIN_FILE_SIZE = 10_000        # 10 KB
MAX_FILE_SIZE = 10_485_760    # 10 MB


class ImageValidationError(Exception):
    """Raised when image validation fails"""
    pass


def validate_image_data(image_data: bytes) -> None:
    """
    Validate image data before processing

    Args:
        image_data: Raw image bytes

    Raises:
        ImageValidationError: If validation fails
    """

    # Check if data exists
    if not image_data:
        raise ImageValidationError("No image data provided")

    # Check file size
    file_size = len(image_data)

    if file_size < MIN_FILE_SIZE:
        raise ImageValidationError(
            f"File too small ({file_size} bytes). Minimum size is {MIN_FILE_SIZE} bytes."
        )

    if file_size > MAX_FILE_SIZE:
        raise ImageValidationError(
            f"File too large ({file_size / 1024 / 1024:.1f} MB). Maximum size is 10 MB."
        )

    # Check for valid image file signatures (magic bytes)
    if not is_valid_image_format(image_data):
        raise ImageValidationError(
            "Unsupported file format. Please upload PNG, JPG, or PDF."
        )

    logger.info(f"Image validation passed: {file_size / 1024:.1f} KB")


def is_valid_image_format(image_data: bytes) -> bool:
    """
    Check if image data has valid file signature (magic bytes)

    Args:
        image_data: Raw image bytes

    Returns:
        True if format is supported (PNG, JPG, PDF)
    """

    if len(image_data) < 8:
        return False

    # PNG signature
    if image_data[:8] == b'\x89PNG\r\n\x1a\n':
        return True

    # JPEG signature
    if image_data[:2] == b'\xff\xd8':
        return True

    # PDF signature
    if image_data[:4] == b'%PDF':
        return True

    return False
